lib/Tumblr/StreamBuilder/StreamRankers/WeightedRandomStreamRanker.php

```diff
@@ -51,8 +51,12 @@ protected function rank_inner(array $stream_elements, StreamTracer $tracer = nul
         foreach ($valid_elements as $element) {
             /** @var RecommendationLeafStreamElementTrait $original_element */
             $original_element = $element->get_original_element();
-            // calculate sampling score
-            $r = pow(mt_rand() / $max_rand, (1 / $original_element->get_score()));
+            if ($original_element->get_score() === 0.0) {
+                $r = 0;
+            } else {
+                // calculate sampling score
+                $r = pow(mt_rand() / $max_rand, (1 / $original_element->get_score()));
+            }
             $H[strval($r)] = $element;
```

Comment:
I don't know if we need to be worried about this line though: `$H[strval($r)] = $element;`. An element with a score of zero should be a rare occurrence, but is it possible some elements will be lost? It seems that way?

Reply:
i feel like we should have some logging somewhere to show us definitively whether a score of `0` actually happens... maybe we should add some here, if not?

Reply:
ok, i see that we do have this logging, since the exception is happening in production. in that case, maybe a decision could be made here to just drop the element entirely if the weight ends up being 0. that seems to make sense to me? cc @nicola-barbieri @lucila  i guess it's a separate broader question of what we should do here when two elements have the exact same score, because yeah, they'll collide here.

Reply:
it might still be possible for two elements to have the same score, even if not zero though

Reply:
When score->0 then r->infinity, in which case I think assigning zero to it is far from the actual final score.  

It all depends on the requirements of this ranker and what a score=0 means: 
1. if it's an error it means nothing and we can give it indeed a very low score or even make this element an invalid one (you can do this below)
2. If it's not an error and it's an acceptable value which is just not handled then `r` should be a high number and not zero or invalid.

Reply:
> just drop the element entirely if the weight ends up being 0

I would not drop any elements because of this. Some sources don't have a score, and it would be likely that the default score is 0.

Reply:
@lovemeblender `mt_rand() <= $max_rand`, so score->0 means r->0. Unless `mt_rand() == $max_rand`, where all elements will get the same score, which seems like yet another bug? mt_rand() should ideally always be less than $max_rand.

Reply:
so what if, theoretically, we get two elements with the same score. Only one of those will end up in `$H`. Does this mean the other one gets discarded?

Reply:
another alternative I'm thinking about, to reduce the chances of having more than 1 element with the same score, is to do something like:

```
$score = $original_element->get_score();
if ($score === 0) {
  $score = 0.001;
}
$r = pow(mt_rand() / $max_rand, (1 / $score));
```
so that we will still call mt_rand() with a very small value on the exp part.

thoughts? @eltongo 

I think this is an alternative @lovemeblender  had suggested before (internally on a huddle)

Reply:
I would think it is also possible for two elements to have the same non-zero score, no? In that case, what we need is to fix `$H`, because this would only fix the case for the score being zero.

Reply:
> mt_rand() <= $max_rand 
if that's true we're set with either zero or adding an very small number to avoid collisions ^  collisions will be handled gracefully I think as the elements will just appear adjacently

Reply:
I agree with fixing the division by zero, maybe adding an epsilon to all input scores:
$score = $score + $epsilon (0.00001) 

score = 0 shouldn't happen in the first place (unless rounding errors). which element is causing this ?